## Importing the necessary libraries

In [14]:
import tensorflow as tf
import numpy as np
import pandas as pd

## Loading the Dataset

In [15]:
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()

# Normalizing the dataset
x_train = x_train/255.0
x_test = x_test/255.0
x_train=x_train.reshape(60000,28,28,1)
x_test=x_test.reshape(10000,28,28,1)

## Building the Neural Network

In [16]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Conv2D(64,(3,3),activation = 'relu',input_shape = (28,28,1)))
model.add(tf.keras.layers.MaxPool2D(2,2))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model.add(tf.keras.layers.Dense(10, activation = tf.nn.softmax))  # we can also write this as activation = 'softmax'
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
# NN has 2 hidden layers   

## Defining a Function to train the model
##  

Here, we have defined a function, which takes the data and model as inputs and returns the loss and accuracy as the output.

The model consists of 3 main steps :
- Defining the callback class and instantiating it to stop training and prevent overfitting
- Fitting the model on the training set
- Evaluating the model's performance on the test set and calculating accuracy

In [17]:
def train_model(x_train,y_train,x_test,y_test,model):
    '''
    function which trains the model and returns the number of epochs as welll as accuracy 
    '''
    class Call_back(tf.keras.callbacks.Callback):
        def on_epoch_end(self,epoch,logs={}):
            if(logs.get('accuracy')>0.99):
                print('Desired Accuracy Reached.Stopping the Training.')
                self.model.stop_training = True
    
    callback = Call_back()
                

    
    # model fitting
    history = model.fit(x_train,y_train,epochs =10,callbacks = [callback])
    metrics = model.evaluate(x_test,y_test)
    return metrics

#### Now, we call the train model function...

In [18]:
train_model(x_train,y_train,x_test,y_test,model)

Train on 60000 samples
Epoch 1/10
60000/60000 [==============================] - 71s 1ms/sample - loss: 0.1348 - accuracy: 0.9594
Epoch 2/10
60000/60000 [==============================] - 69s 1ms/sample - loss: 0.0479 - accuracy: 0.9850
Epoch 3/10
10000/10000 [==============================] - 3s 347us/sample - loss: 0.0492 - accuracy: 0.9854


[0.04919529082388617, 0.9854]

### As we can see we have a loss of 0.08 and an accuracy of 98% on the validation set 